### ISSUE TITLE
Labels and Selectors - a common practice within the templates

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/zero-to-jupyterhub-k8s#538]
#### Related issue: #615

---

I was about to make set labels and selectors for a StatefulSet I tried out, and wanted to mimic what had been done for other k8s object templates, but did not find a clear pattern.

So, I'd like to figure out a practice to follow within the guide.
- What labels do we declare?
- What are important labels within a selector?

```yaml
metadata:
  name: proxy
  labels:
    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
    component: proxy
    heritage: {{ .Release.Service }}
    release: {{ .Release.Name }}
spec:
  replicas: 1
  selector:
    matchLabels:
      component: proxy
      release: {{ .Release.Name }}
      heritage: {{ .Release.Service }}
```

I'd like to discuss the following labels:
- chart, app, component, heritage, release

### My take
- I recommend we don't use the `chart` label, or is there a benefit to including that?
- I recommend we select on few labels in a minimalistic manner, perhaps only component or component and release is needed? What can go wrong by including / not including various labels?

Perhaps something like below?

#### The labels we declare on controllers
- [x] chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
- [x] app: jupyterhub
- [x] component: hub
- [x] release: {{ .Release.Name }}
- [x] heritage: {{ .Release.Service }}

#### The important labels within a selector and a controllers pod template
- [ ] chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
- [x] app: jupyterhub
- [x] component: hub
- [x] release: {{ .Release.Name }}
- [ ] heritage: {{ .Release.Service }}

**Comment by yuvipanda** [comment]
https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/labels.md is where I took my cues from, and I think we should stick to it. Note that I didn't see that page until after a few releases, so it isn't consistently applied.

I'm happy for us to standardize this! I think we should include all the 5 ones that helm recommends in all our components. There's nothing wrong with changing them, but I think sticking to the standards helps other tooling / people who are used to helm charts to identify / write additional scripts that deal with z2jh.

**Comment by consideRatio** [comment]
Ah! What about label selection in a `selector:`? It would be overkill to select based on every label. What about going for component + release?

**Comment by yuvipanda** [comment]
@consideRatio good question! I wonder what the charts in github.com/kubernetes/charts do. We should probably follow whatever it is they do...

**Comment by gsemet** [comment]
Hello. I like the fact that every names are prepended with the release name, so that two helm deployment can be done on the same namespace without conflict. Do you think it can be feasible for Zero-to-Jupyterhub?
In `_helpers.tpl` we would have:
```
{{/*
Create a default fully qualified app name.
We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
*/}}
{{- define "jupyterhub.fullname" -}}
{{- printf "%s" .Release.Name | trunc 63 | trimSuffix "-" -}}
{{- end -}}
```
And all name metadata are prepended with this:
- Configmap:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: "{{ template "jupyterhub.fullname" . }}-env"
data:
```
- Deployment:
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: {{ template "jupyterhub.fullname" . }}-myapp
```
And so on.

**Comment by minrk** [comment]
@gsemet while it's not impossible for two hubs in one namespace to work, I think it's probably good practice to always deploy only one jupyterhub per namespace. Namespaces are a very low-cost entity, and it makes it very easy to be sure what belongs together.

There's one case where we have [`app: kube-lego`](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/e7a147d51d635e4be413ec9ce2bf9d6d1455fccd/jupyterhub/templates/proxy/deployment.yaml#L31-L32) which has a comment stating that this is needed for kube-lego to do its magic. Maybe there's another way? I do think we should have `app: jupyterhub` on all of these, if we can.

**Comment by gsemet** [comment]
I understand how I do not have this option on our environment (baremetal with team that is ramping up on this subject). For automatic test, it may be useful to use several deployment even in the same namespace. Helm charts usually add the "release name" in front of each element (pod, configmap, claim,...) to ease the readability, I find it a good practice.

**Comment by consideRatio** [comment]
@minrk the `app: kube-lego` situation is will go away by #592 !

@gsemet thanks for insights into this! For that reason it makes sense to me to add a release to the name. I will probably clutter output from `kubectl get pod` etc and make the need for a kubectl autocomplete even bigger though.

I conclude that I don't understand enough about namespaces to be confident enough on a stance for or against including release in the kubernetes objects' names or not.

#### Related
What is a release? [The helm documentation.](https://github.com/kubernetes/helm/blob/master/docs/glossary.md#release)

**Comment by consideRatio** [comment]
I'll make a PR letting the name field (not a label) be as it is, and apply chart, app, component, release, and heritage labels, and systematically use a simple selector only selecting on component.

If we decide on updating the name to include the helm release name, we should also select based on the release label.

**Comment by minrk** [comment]
@consideRatio that sounds great!

**Comment by consideRatio** [comment]
@minrk got any ideas on the chart label? ingress.yaml is deviating from the others for some reason I fail to find, but I recall reading something about this, perhaps it was in the gitter chat.

https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/be3d3b792c9373643bb72fcdf9d4b22c67a5f247/jupyterhub/templates/ingress.yaml#L8

Should I make all chart labels replace +, only the ingress.yaml one, or none?

**Comment by yuvipanda** [comment]
@gsemet The bigger problem with multiple hubs in a namespace is that kubespawner doesn't support that atm - it assumes it 'owns' the namespace. We need to fix that first IMO.

Changing the names might also have backwards compatibility issues, primarily around renaming the hub PVC. So we should do that carefully and IMO make a point release specifically for it.

**Comment by yuvipanda** [comment]
@consideRatio I think that .replace exists simply from copy pasting from kubernetes/charts :) I think including it is a good option!

**Comment by gsemet** [comment]
I know that helm support hooks that may be used (I guess) to handle smooth transition, maybe it can do the pod rename... never tried it myself sadly

**Comment by consideRatio** [comment]
@yuvipanda according to [helm's best practices](https://docs.helm.sh/chart_best_practices/#standard-labels) one should use the replace + with _. Still not confident why. I also found about 1/4 of all the kubernetes/charts to do that.

@gsemet regarding the name field, it seems almost all charts in kubernetes/charts is naming their objects similarly to how you suggest! I'm creating an issue for further discussion about implementing helm-release dependent naming.

**Comment by gsemet** [comment]
i guess restriction related to chart templating has something to do with DNS, when one use systems such as calico or other advanced networking environment.

**Comment by consideRatio** [comment]
Closed by #625
